### DISCUSSION TITLE
Clarification on intended usage

### DISCUSSION THREAD

**Discussion by AdamDiament** [mattleibow/DeviceRunners#17] [Q&A] [answered]
Hey! My name's Adam, I'm from Jerusalem. I'm in the midst of upgrading a large XF project to MAUI. We have a UI test suite but because the UI tests take so long to run, we've been wanting to get a business logic view model unit test suite built to supplement the UI tests. These would just require instantiating viewmodels and running logic tests, rather than booting up the whole app and needing a device or emulator etc. The recommended approach towards getting business logic tests running in XF (see the XF example [here](https://learn.microsoft.com/en-us/xamarin/xamarin-forms/enterprise-application-patterns/unit-testing)) required developing an extra test runner app, per platform, which is an extra layer of complication I could do without. So this project looks really cool.

I just wanted to check - would the idea be that this would be integrated inside the app itself, or to create a separate app just for running tests as per the old way. From what I can see from your code, the tests can run from within the app itself. Which is very cool if I've understood correctly. Thanks.

**Comment by mattleibow** [comment, marked as answer]
Hi Adam! Good to hear that you are upgrading, hope it is going well so far.

With regards to testing, there are a few options and we use all of them at once in the maui repo. I started writing a replay but it got so long so I decided to make it a wiki for all time: https://github.com/mattleibow/DeviceRunners/wiki/Types-of-Tests

**Reply by AdamDiament** [reply]
Hey thanks for the reply @mattleibow. The MAUI upgrade is going ok, but I've hit nuget related blockers around upgrading my native binding to the Zoom Meeting SDK (https://github.com/VisualService/dotnet-zoom-meeting-SDK/issues/11) - but that's another story haha. 

The wiki looks really useful thanks. Fundamentally I think what you're saying is that to do `type 2) On-Device Testing (using DeviceRunners)` I should expose my viewmodels and views via a project reference to the device runner app, and the test runner app will register and load the page inside itself to perform the test? I can mock the service dependencies of the viewmodel etc, but wouldn't the test runner app need access to all the resources in my app too, to load to xaml page properly? Sorry if I'm fundamentally misunderstanding something here!

(In your example, the device runner app has access to both MonkeysPage and ViewModel. Presumably defined in the app to be tested)

**Reply by mattleibow** [reply]
I created this issue so we can have a look at this: https://github.com/mattleibow/DeviceRunners/issues/19

Very good point. In my scenarios, the view and view model code was in a separate class library so test code can be used. Like the page to be shown but with test services. Running tests on a server that has latency and require auth is not the best.
